k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -28,25 +27,23 @@ class EndpointsNamer(
     case id@Path.Utf8(nsName, portName, serviceName) =>
       val residual = path.drop(PrefixLen)
       log.debug("k8s lookup: %s %s", id.show, path.show)
-      Ns.get(nsName).services.flatMap { services =>
-        log.debug("k8s ns %s initial state: %s", nsName, services.keys.mkString(", "))
-        services.get(serviceName) match {
+      Ns.get(nsName).map { nsCache =>
+        log.debug("k8s ns %s initial state: %s", nsName, nsCache.services.keys.mkString(", "))
+        nsCache.services.get(serviceName) match {
           case None =>
             log.debug("k8s ns %s service %s missing", nsName, serviceName)
-            Activity.value(NameTree.Neg)
+            NameTree.Neg
 
           case Some(services) =>
             log.debug("k8s ns %s service %s found", nsName, serviceName)
-            services.ports.map { ports =>
-              ports.get(portName) match {
-                case None =>
-                  log.debug("k8s ns %s service %s port %s missing", nsName, serviceName, portName)
-                  NameTree.Neg
+            services.ports.get(portName) match {
+              case None =>
+                log.debug("k8s ns %s service %s port %s missing", nsName, serviceName, portName)
+                NameTree.Neg
 
-                case Some(port) =>
-                  log.debug("k8s ns %s service %s port %s found + %s", nsName, serviceName, portName, residual.show)
-                  NameTree.Leaf(Name.Bound(port.addr, idPrefix ++ id, residual))
-              }
+              case Some(port) =>
+                log.debug("k8s ns %s service %s port %s found + %s", nsName, serviceName, portName, residual.show)
+                NameTree.Leaf(Name.Bound(Var(port.addr), idPrefix ++ id, residual))
```

Comment:
1. Name.Bound is not required to be a path (it may be a reference, for instance).
2. Multiple NameTrees may return the same Name.Bound -- "same" being determined by id equality.
3. clients are cached on this.   Load balancers observe updates to the Var[Addr] and update their list of endpoints accordingly.

It's probably helpful to review finagle's client stack (ignoring linkerd for the time being)--specifically, BindingFactory and how it's layered onto LoadBalancerFactory.